ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/DeprecatedStatementsCheckerTest.java

```diff
@@ -0,0 +1,357 @@
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import com.google.common.collect.Iterables;
+import io.confluent.ksql.execution.ddl.commands.KsqlTopic;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.metastore.TypeRegistry;
+import io.confluent.ksql.metastore.model.DataSource;
+import io.confluent.ksql.metastore.model.KsqlStream;
+import io.confluent.ksql.metastore.model.KsqlTable;
+import io.confluent.ksql.name.SourceName;
+import io.confluent.ksql.parser.AstBuilder;
+import io.confluent.ksql.parser.DefaultKsqlParser;
+import io.confluent.ksql.parser.KsqlParser;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+import java.util.Optional;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DeprecatedStatementsCheckerTest {
+  private final KsqlParser STATEMENT_PARSER = new DefaultKsqlParser();
+
+  private final SourceName STREAM_1 = SourceName.of("STREAM_1");
+  private final SourceName STREAM_2 = SourceName.of("STREAM_2");
+  private final SourceName STREAM_3 = SourceName.of("STREAM_3");
+  private final SourceName TABLE_1 = SourceName.of("TABLE_1");
+
+  private DeprecatedStatementsChecker statementsChecker;
+
+  @Mock
+  private MetaStore metaStore;
+
+  @Before
+  public void setup() {
+    statementsChecker = new DeprecatedStatementsChecker(metaStore);
+
+    final DataSource streamSource1 = newKsqlStream(STREAM_1);
+    when(metaStore.getSource(STREAM_1)).thenReturn(streamSource1);
+
+    final DataSource streamSource2 = newKsqlStream(STREAM_2);
+    when(metaStore.getSource(STREAM_2)).thenReturn(streamSource2);
+
+    final DataSource streamSource3 = newKsqlStream(STREAM_3);
+    when(metaStore.getSource(STREAM_3)).thenReturn(streamSource3);
+
+    final DataSource tableSource1 = newKsqlTable(TABLE_1);
+    when(metaStore.getSource(TABLE_1)).thenReturn(tableSource1);
+  }
+
+  @Test
+  public void shouldDeprecateStreamStreamLeftJoinWithNoGrace() {
+    // Given
+    final Statement statement = parse(String.format(
+            "SELECT * FROM %s AS l "
+            + "LEFT JOIN %s AS r WITHIN 1 SECOND ON l.K = r.K;",
+        STREAM_1.text(),
+        STREAM_2.text()
+    ));
+
+    // When
+    final Optional<DeprecatedStatementsChecker.Deprecations> deprecations =
+        statementsChecker.checkStatement(statement);
+
+    // Then
+    assertThat(deprecations, is(Optional.of(DEPRECATED_STREAM_STREAM_OUTER_JOIN_WITH_NO_GRACE)));
+  }
+
+  @Test
+  public void shouldDeprecateStreamStreamLeftJoinWithNoGraceOnMultiStreamStreamJoin() {
```

Comment:
Can we replicate this test with different join order? We should be able to detect both cases and warn for each.